### DISCUSSION TITLE
Sun and Moon rise and set times. Moon phase.

### DISCUSSION THREAD

**Discussion by peterhinch** [micropython/micropython#13189] [Show and tell]
[This module](https://github.com/peterhinch/micropython-samples/tree/master/astronomy) is a port of some C code from a textbook, with adaptation for global use. It enables an application to determine the times of rise and set events in a given day. It can be configured for any location on the globe. It can be used alone, or as part of the [schedule module](https://github.com/peterhinch/micropython-async/blob/master/v3/docs/SCHEDULE.md). Using `schedule` allows you to set up commands like "turn on the lights at one hour after sunset, but only in the winter months". Or whatever.

Another possible application is in clocks such as the one below. I built this before MicroPython was written, but it could easily be done in MP.

![Image](https://github.com/peterhinch/micropython-samples/blob/master/astronomy/lunartick.jpg?raw=true)

**Comment by rkompass** [comment]
Very nice painting.

**Reply by peterhinch** [reply]
It is good, but I can't take the credit: Sue is the artist :)

**Comment by shariltumin** [comment]
Nice!!

**Comment by scruss** [comment]
I like this, and really appreciate your efforts to keep it accurate. Now all I need to do is try to understand asyncio

Any chance that it could be extended to return twilight times? Civil twilight is when the sun is 6° below the horizon, and is useful for automatic lighting.

**Reply by peterhinch** [reply]
I'll consider it.

**Reply by peterhinch** [reply]
I've pushed an update which supports the various definitions of twilight. A constructor arg allows you to specify a number of degrees below the horizon for twilight, so a `RiSet` instance can support one twilight definition. The default (`None`) saves processor time by skipping twilight calculation.

**Reply by scruss** [reply]
Thank you! Works a treat:
```python3
from sched.sun_moon import RiSet

# Civil twilight is 6° below horizon
toronto = RiSet(lat=43.729, long=-79.292, lto=-5, tl=6)
print(f"Sun rise {toronto.sunrise(3)} set {toronto.sunset(3)}")
print(f"Civil twilight start {toronto.tstart(3)} end {toronto.tend(3)}")
```
Output for today:
```
Sun rise 07:44:27 set 16:41:07
Civil twilight start 17:13:33 end 07:11:32
```
Compares well to Dan Risacher's [sunwait](https://risacher.org/sunwait/):
```
sunwait -p 43.729N 79.292W
Using location:             43.729000N, 79.292000W
Date:                       16 Dec 2023 
Local time:                 12:11 
 …
Current specified time zone: EST (-5 from UTC) 
Sun transits meridian 1213 EST
                   Sun rises 0745 EST, sets 1641 EST
       Civil twilight starts 0712 EST, ends 1714 EST
```
You might have twilight start/end swapped: it's always twilight when the sun's up.

**Reply by peterhinch** [reply]
Tanks or testing.

> You might have twilight start/end swapped

Indeed, will fix.
